apps/algorithm/search/src/server.py

```diff
@@ -65,35 +129,99 @@ async def search(
     limit: int = 20,
     minPrice: float = None,
     maxPrice: float = None,
-    status: str = "AVAILABLE",
-    searchType: str = "LISTINGS",
-    sort: str = "RELEVANCE",
+    status: Status = "AVAILABLE",
+    searchType: SearchType = "LISTINGS",
+    sort: Sort = "RELEVANCE",
 ):
-    # actual logic will go here
-
-    # This is a dummy response
-    return [
-        {
-            "listingID": "A23F29039B23",
-            "sellerID": "A23F29039B23",
-            "sellerName": "John Doe",
-            "title": "Used Calculus Textbook",
-            "description": "No wear and tear, drop-off available.",
-            "price": 50,
-            "dateCreated": "2024-05-23T15:30:00Z",
-            "imageUrl": "image URL for first Image",
-        },
-        {
-            "listingID": "A23F29039B24",
-            "sellerID": "A23F29039B24",
-            "sellerName": "Jane Doe",
-            "title": "Used Physics Textbook",
-            "description": "No wear and tear, drop-off available.",
-            "price": 40,
-            "dateCreated": "2024-05-23T15:30:00Z",
-            "imageUrl": "image URL for first Image",
-        },
-    ]
+    try:
+        validate_search_params(latitude, longitude, page, limit, minPrice, maxPrice)
+
+        INDEX = os.getenv("ES_INDEX", DEFAULT_INDEX)
+
+        sort_options = {
+            "RELEVANCE": "_score",
+            "PRICE_ASC": "price",
+            "PRICE_DESC": "price",
+            "LISTED_TIME_ASC": "dateCreated",
+            "LISTED_TIME_DESC": "dateCreated",
+            "DISTANCE_ASC": "_geo_distance",
+            "DISTANCE_DESC": "_geo_distance",
+        }
+
+        if searchType == "LISTINGS":
+            must_conditions = [
+                {"multi_match": {"query": query, "fields": ["title", "description"]}},
+                {"match": {"status": status}},
+            ]
+        elif searchType == "USERS":
+            must_conditions = [
+                {"match": {"sellerName": query}},
+                {"match": {"status": status}},
+            ]
+        else:
+            raise HTTPException(status_code=400, detail="Invalid searchType")
```

Comment:
error codes should be consistent, generally 400 is used for a missing or incorrect parameter, and 422 is used for a syntactically correct parameter that fails business rules. In this case there's a valid searchType field with a bad value, so it should return 422

Reply:
did it